test-app/tests/unit/system/upload-progress-test.ts

```diff
@@ -0,0 +1,202 @@
+/* eslint-disable qunit/no-conditional-assertions */
```

Comment:
Not sure if there's a good way to avoid this 🤔 

The data structure and complexity is a bit high for `assert.step` and `assert.verifySteps`.

I'm pretty confident that `replayEvents` is simple enough that assertions are running when we expect them to.